const Dotenv = require('dotenv');
const Express = require('express');
const Mongoose = require('mongoose');
const Cors = require('cors');

require('express-async-errors');

Dotenv.config();

const DatabaseConfig = require('../config/database');
const Routes = require('./routes');
const AppError = require('./errors/AppError');

class App {
  constructor() {
    this.server = Express();

    this.database();
    this.middwares();
    this.routes();
    this.exception();
  }

  database() {
    Mongoose.connect(DatabaseConfig.uri, {
      useCreateIndex: true,
      useNewUrlParser: true,
      useFindAndModify: false,
      useUnifiedTopology: true,
    });
  }

  middwares() {
    this.server.use(Cors());
    this.server.use(Express.json());
    this.server.use(Express.urlencoded({ extended: true }));
  }

  routes() {
    this.server.use(Routes);
  }

  exception() {
    this.server.use(async (err, req, res, next) => {
      if (err instanceof AppError) {
        return res.status(err.statusCode).json(err.error);
      }

      return res.status(500).json({
        status: 'error',
        message: 'Internal Server Error',
      });
    });
  }
}

module.exports = new App().server;
